src/metabase/pulse/markdown.clj

```diff
@@ -0,0 +1,320 @@
+
+(def ^:private parser
+  "An instance of a Flexmark parser"
+  (delay (.build (Parser/builder))))
+
+(def ^:private node-to-tag-mapping
+  "Mappings from Flexmark AST nodes to keyword tags"
+  {Document              :document
+   Paragraph             :paragraph
+   ThematicBreak         :horizontal-line
+   HardLineBreak         :hard-line-break
+   SoftLineBreak         :soft-line-break
+   Heading               :heading
+   StrongEmphasis        :bold
+   Emphasis              :italic
+   OrderedList           :ordered-list
+   BulletList            :unordered-list
+   OrderedListItem       :ordered-list-item
+   BulletListItem        :unordered-list-item
+   Code                  :code
+   FencedCodeBlock       :codeblock
+   IndentedCodeBlock     :codeblock
+   BlockQuote            :blockquote
+   Link                  :link
+   Reference             :reference
+   LinkRef               :link-ref
+   ImageRef              :image-ref
+   Image                 :image
+   AutoLink              :auto-link
+   MailLink              :mail-link
+   HtmlEntity            :html-entity
+   HtmlBlock             :html-block
+   HtmlInline            :html-inline
+   HtmlCommentBlock      :html-comment-block
+   HtmlInlineBase        :html-inline-base
+   HtmlInlineComment     :html-inline-comment
+   HtmlInnerBlockComment :html-inner-block-comment})
+
+(defn- node-to-tag
+  [node]
+  (node-to-tag-mapping (type node)))
+
+(defprotocol ^:private ASTNode
+  (to-clojure [this _]))
+
+(defn- convert-children [node source]
+  (map #(to-clojure % source) (.getChildren ^Node node)))
+
+(extend-protocol ASTNode
+  Node
+  (to-clojure [this source]
+    {:tag     (node-to-tag this)
+     :attrs   {}
+     :content (convert-children this source)})
+
+  Text
+  (to-clojure [this _]
+    (str (.getChars this)))
+
+  FencedCodeBlock
+  (to-clojure [this source]
+    {:tag     (node-to-tag this)
+     :attrs   {}
+     :content (str (.getContentChars this))})
+
+  IndentedCodeBlock
+  (to-clojure [this _]
+    {:tag     (node-to-tag this)
+     :attrs   {}
+     :content (str (.getContentChars this))})
+
+  Link
+  (to-clojure [this source]
+    {:tag     (node-to-tag this)
+     :attrs   {:href (str (.getUrl this))
+               :title (not-empty (str (.getTitle this)))}
+     :content (convert-children this source)})
+
+  Reference
+  (to-clojure [this _]
+    {:tag   (node-to-tag this)
+     :attrs {:title (not-empty (str (.getTitle this)))
+             :label (str (.getReference this))
+             :url (str (.getUrl this))}})
+
+  LinkRef
+  (to-clojure [this source]
+    {:tag     (node-to-tag this)
+     :attrs   {:reference (-> (.getDocument this)
+                              (.get Parser/REFERENCES)
+                              (get (str/lower-case (str (.getReference this))))
+                              (#(to-clojure % source)))}
```

Comment:
this is equivalent to just `(to-clojure source)` (although i still think `source` should go away)